### ISSUE TITLE
Feature request: a way of loading row matrices

### ISSUE THREAD

**Issue by magcius** [gpuweb/gpuweb#4094]
It's convenient to want to pack a mat4 row with an implicit last column 0,0,0,1 into 12 floats for storage efficiency. However, with our packing/alignment rules, that's a bit difficult to do. The easiest way I can see is to use a custom structure and three floats:

```wgsl
struct Mat4x3 {
   vec4f mx,
   vec4f my,
   vec4f mz,
};

fn unpack_Mat4x3(a: Mat4x3) -> mat4x4f {
  return transpose(mat4x4(a.mx, a.my, a.mz, vec4f(0, 0, 0, 1)));
}
```

which is a bit annoying, especially when handling both loads and stores.

This purpose is similarly served by the `row_major` annotation in GLSL, the pack_matrix pragma in HLSL, and the RowMajor decoration in SPIR-V. I'm not sure exactly what we might want the syntax to be to make it clear that this is only about the packing in the structure and not the operations on the matrix itself, which has been a confusion for GLSL.

btw, I noticed that `column_major` is a reserved word, but `row_major` is not. Was this intentional?

**Comment by dneto0** [comment]
As I recall, the broad consensus was that row-major matrices as a kind of matrix type were a major headache and source of bugs. 

I haven't given much thought to builtin functions in the spirit of packing/unpacking. That's a much smaller footprint for a feature, and worth considering.

**Comment by kdashg** [comment]
Ideally isn't this:
```
fn foo(a: mat3x4) -> mat4x4f {
  return mat4x4(transpose(mat3x4));
}
```

**Comment by kainino0x** [comment]
I don't think it's necessarily clear whether that would fill with 0,0,0,0 or 0,0,0,1.

**Comment by magcius** [comment]
> As I recall, the broad consensus was that row-major matrices as a kind of matrix type were a major headache and source of bugs.

I agree that matrices are confusing, especially because both GLSL and HLSL get them backwards in subtle but different ways. In both GLSL and HLSL, the row_major attribute only affects the storage order when loading from/storing to buffers, but not the runtime behavior. The 99% case for the attributes is to match a native Matrix type so you can use them directly in your data structures, or to store an affine matrix in 3 rows of 4.

> I haven't given much thought to builtin functions in the spirit of packing/unpacking.

Builtin functions for packing/unpacking might be OK. I don't know what the roadmap for WGSL looks like, but the [ByteAddressBuffer typed Load/Stores](https://github.com/microsoft/DirectXShaderCompiler/wiki/ByteAddressBuffer-Load-Store-Additions) are extremely nice to write code with, and something that works well with a feature like that would be appreciated.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-05-23-minutes#feature-request-a-way-of-loading-row-matrices--issue-4094">WGSL 2023-05-23 Minutes</a></summary>

* KG: two issues.  Rowm vs. colm memory access, and getting the last column to inject the 0,0,0,1.  
* KG: Consider using ‘transpose’ for converting.  Second half is the 0,0,0,1.   For that, consider an automatic expansion that fills those values in.
* AB: Seems easy in the spec, and well-defined.  No opinion on how generally useful it is.  Seems similar to vector construction that allows you to give mix of scalars and vectors to make a wider thing.
* KG: The ambiguity here is …
* JB: Seems would be nice to have mat4x4 overloads that added the 0,0,0,1
* **KG: What about getting a 3x4 from a 4x4.  **
* **JB: That loses information. Let’s not do that.**
* DN: Would be different function; a projection.
* JB: A differently named thing: homogenous value from affine?
* AB: Function to construct mat4x4 by providing 4x3 with an extra vector.
* KG: Clunkiness comes from hard to make a (0,0,0,1)
* DS: If you create a default matrix it’s all zeros, but now it extends with not-zeros.
* MM: Better to add builtin functions than to add row-major attributes to data…..
* KG: **No on attributes for row-major.**  Should we reserve row_major?
* MM: Seems we got it wrong. **Should have reserved row_major.**
* MM: Sounds like direction: no attribute.  But **proposals welcome for new std library functions**.
* DN: Yes.

(Followup.  Reviewing history, I found that row_major was dropped as a reserved word when we stopped reserving type names.  See [https://github.com/gpuweb/gpuweb/pull/3167](https://github.com/gpuweb/gpuweb/pull/3167)  

**Maybe column_major should have been removed also**?)

</details>
